### PR TITLE
Product Search

### LINKED ISSUE #1285: Create a Mock Search API to provide result data to pass to the UI (v1)
As a first step in implementing search, we should mock up an API that will stub / return (not sure if those are the right terms) data to test with the UI portion of search.

### PR DESCRIPTION
**Pull request by zenweasel**
Resolves #1285 #1287 
It's not actually "mock" anymore, this is the real one.

backend is in "search-mongo" plugin
frontend is in "ui-search"

### CODE REVIEW

File: imports/plugins/included/search-mongo/server/startup/init.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+Hooks.Events.add("afterCoreInit", () => {
+  // for (let i = 1; i < 200; i++) {
+  //   Logger.info("Inserting product: ", i);
+  //   createProduct();
+  // }
+  const existingDoc = Products.find().count();
+  if (existingDoc === 4) {
+    Logger.warn("======> Adding Search Products");
+    // noinspection JSUnresolvedVariable
+    const tagsJson = Assets.getText("custom/SearchTags.json");
```
Comment (aaronjudd) [inline imports/plugins/included/search-mongo/server/startup/init.js]: `Error: Unknown asset: custom/SearchTags.json`

Comment (zenweasel) [inline imports/plugins/included/search-mongo/server/startup/init.js]: You have to copy the files from the `data` directory to `private/custom`. There's a README there but I should probably put it in the PR notes.

Comment (aaronjudd) [inline imports/plugins/included/search-mongo/server/startup/init.js]: Just noting that the error is thrown when you do a fresh install of Reaction, and that this error should be handled.

Comment (zenweasel) [inline imports/plugins/included/search-mongo/server/startup/init.js]: Everything in `startup` is only for @kieckhafer and I to use for testing and will not be in the final PR.

File: imports/plugins/included/ui-search/client/templates/searchModal.html [hunk 2]
```diff
@@ -15,7 +15,7 @@
 
       {{#if tagSearchResults}}
       <div class="rui search-modal-tags-container">
-        <p class="suggested-tags">Suggested tags</p>
+        <p class="rui suggested-tags">Suggested tags</p>
```
Comment (aaronjudd) [inline imports/plugins/included/ui-search/client/templates/searchModal.html]: Needs i18n.

File: imports/plugins/included/search-mongo/client/settings/search.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    Reaction.showActionView();
+  }
+});
+
+AutoForm.hooks({
+  "search-update-form": {
+    /* eslint-disable no-unused-vars*/
+    onSuccess() {
+      Alerts.removeSeen();
+      return Alerts.add("Search settings saved.", "success");
```
Comment (aaronjudd) [inline imports/plugins/included/search-mongo/client/settings/search.js]: Needs i18n.

File: imports/plugins/included/search-mongo/server/methods/searchcollections.js [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+import { Reaction, Logger } from "/server/api";
+import { ProductSearch, OrderSearch, AccountSearch, Orders, Products, Accounts, Shops } from "/lib/collections";
+import utils from "./common";
+import { transformations } from "./transformations";
+
+
+const requiredFields = {};
+requiredFields.products = ["_id", "hashtags", "shopId", "handle", "price", "isVisible"];
+
+const supportedLanguages = ["da", "nl", "en", "fi", "fr", "de", "hu", "it", "nb", "pt", "ro", "ru", "es", "sv", "tr"];
```
Comment (aaronjudd) [inline imports/plugins/included/search-mongo/server/methods/searchcollections.js]: Can we add some comments here about this `supportedLanguages` functionality?

**Comment by kieckhafer** [comment]
i18n is in the process of being added, will push to this PR when ready.

**Comment by mikemurray** [comment]
👍  for pullapprove's sake

**Comment by aaronjudd** [comment]
Fresh install ("rrr"), immediate test of search (with "test") is giving me:

```
I20160916-10:52:40.508(-7)? Exception from sub SearchResults id Qm4bpXZ7Zc3cDdC7P Error: Exception while polling query {"collectionName":"ProductSearch","selector":{"shopId":"J8Bhq3uTtdgwZx3rz","$text":{"$search":"te"},"isVisible":true},"options":{"transform":null,"fields":{"score":{"$meta":"textScore"},"title":1,"hashtags":1,"description":1,"handle":1,"price":1},"sort":{"score":{"$meta":"textScore"}},"limit":99}}: text index required for $text query
I20160916-10:52:40.508(-7)?     at Object.Future.wait (/Users/aaronjudd/.meteor/packages/meteor-tool/.1.4.1_1.ivuj7v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:449:15)
I20160916-10:52:40.508(-7)?     at ObserveMultiplexer.addHandleAndSendInitialAdds (packages/mongo/observe_multiplex.js:55:23)
I20160916-10:52:40.509(-7)?     at MongoConnection._observeChanges (packages/mongo/mongo_driver.js:1271:15)
I20160916-10:52:40.509(-7)?     at Cursor.observeChanges (packages/mongo/mongo_driver.js:946:22)
I20160916-10:52:40.509(-7)?     at Function.Mongo.Collection._publishCursor (packages/mongo/collection.js:344:30)
I20160916-10:52:40.509(-7)?     at Cursor._publishCursor (packages/mongo/mongo_driver.js:927:27)
I20160916-10:52:40.509(-7)?     at Subscription._publishHandlerResult (packages/ddp-server/livedata_server.js:1068:13)
I20160916-10:52:40.510(-7)?     at Subscription._runHandler (packages/ddp-server/livedata_server.js:1041:10)
I20160916-10:52:40.510(-7)?     at Session._startSubscription (packages/ddp-server/livedata_server.js:845:9)
I20160916-10:52:40.510(-7)?     at Session.sub (packages/ddp-server/livedata_server.js:617:12)
I20160916-10:52:40.510(-7)?     at packages/ddp-server/livedata_server.js:551:43
I20160916-10:52:40.510(-7)?     - - - - -
I20160916-10:52:40.511(-7)?     at PollingObserveDriver._pollMongo (packages/mongo/polling_observe_driver.js:161:11)
I20160916-10:52:40.511(-7)?     at Object.task (packages/mongo/polling_observe_driver.js:90:12)
I20160916-10:52:40.511(-7)?     at [object Object]._.extend._run (packages/meteor/fiber_helpers.js:147:1)
I20160916-10:52:40.511(-7)?     at packages/meteor/fiber_helpers.js:125:1```

**Comment by kieckhafer** [comment]
Immediate test without adding any products?

**Comment by kieckhafer** [comment]
@aaronjudd do you have the rest of the console.log from that session?

It seems to take about 10 seconds between the app initializing with no ProductSearch records, and ProductSearch records being created. My thought is if you searched inside of this time period, your error might occur.

````
I20160916-11:13:24.760(-7)? 18:13:24.615Z  INFO Reaction: No ProductSearch records found. Adding build ProductSearch Collection to jobs
I20160916-11:13:24.760(-7)? 18:13:24.639Z  INFO Reaction: Reaction.init() has run
I20160916-11:13:24.760(-7)? 18:13:24.639Z  INFO Reaction: Reaction initialization finished.
=> Started your app.

=> App running at: http://localhost:3000/
I20160916-11:13:35.033(-7)? 18:13:35.032Z  INFO Reaction: ProductSearch collection (re)built successfully.
````

I am trying to reload the app immediately after an rrr, but before the `ProductSearch collection (re)built successfully.` happens, but my machine is too slow, the ProductSearch collection is rebuilt before the app loads.

**Comment by aaronjudd** [comment]
@kieckhafer yup, it's fine if tested after a few seconds when `ProductSearch collection (re)built successfully.` appears in the console.  I think it's maybe an edge case that's ok to ignore for now, but wish it was handled more elegantly.

**Comment by aaronjudd** [comment]
👍  approved merging.

**Pull request merged**